spring-integration-core/src/main/java/org/springframework/integration/handler/AbstractMessageProducingHandler.java

```diff
@@ -150,8 +167,30 @@ else if (reply instanceof AbstractIntegrationMessageBuilder) {
 			}
 		}
 
-		Message<?> replyMessage = createOutputMessage(reply, requestHeaders);
-		sendOutput(replyMessage, replyChannel);
+		if (this.asyncReplySupported && reply instanceof ListenableFuture<?>) {
+			ListenableFuture<?> future = (ListenableFuture<?>) reply;
+			final Object fallbackReplyChannel = replyChannel;
+			future.addCallback(new ListenableFutureCallback<Object>() {
+
+				@Override
+				public void onSuccess(Object result) {
+					sendOutput(createOutputMessage(result, requestHeaders), fallbackReplyChannel);
+				}
+
+				@Override
+				public void onFailure(Throwable ex) {
+					Throwable result = ex;
+					if (!(ex instanceof MessagingException)) {
+						result = new MessageHandlingException(requestMessage, ex);
+					}
+					sendOutput(createOutputMessage(result, requestHeaders), fallbackReplyChannel);
```

Comment:
No, my point was about the explicit `outputChannel`. See the `sendOutput`:
```
MessageChannel outputChannel = getOutputChannel();
if (outputChannel != null) {
	replyChannel = outputChannel;
}
```
So, this error is sent to the donwstream.
OI just mean that we always must send error to the `errorChannel` from headers.
For this purpose the upstream `<header-enricher>` makes sense, if you are going to make such an async interaction in the next `<service-activator>`.